google-analytics-dashboard.html

```diff
@@ -32,66 +32,126 @@
+       * @type object
        */
-      queryUpdated: function(event) {
-        // Update `this.query` with the passed event data.
-        Object.keys(event.detail).forEach(function(key) {
-          this.query[key] = event.detail[key];
-        }.bind(this))
+      query: {
+        type: Object,
+        value: function() { return {} }
+      },
 
-        this.updateChildren();
+      authorized: {
+        type: Boolean,
+        value: false,
+        reflectToAttribute: true
       },
 
-      /**
-       * The `updateChildren` method updates each of this dashboards
-       * `<google-analytics-chart>` element with its current query value.
-       *
-       * @method updateChildren
-       */
-      updateChildren: function() {
-        var charts = this.querySelectorAll('google-analytics-chart');
-        for (var i = 0, chart; chart = charts[i]; i++) {
-          Object.keys(this.query).forEach(function(key) {
-            chart[key] = this.query[key];
-          }.bind(this));
-        }
+      errorMessage: {
+        type: String,
+        value: ''
+      },
+
+      hasError: {
+        type: Boolean,
+        computed: 'computeHasError(errorMessage)'
+      }
+    },
+
+    listeners: {
+      'analytics-dashboard-control-change': 'queryUpdated',
+      'analytics-query-error': 'handleQueryError'
+    },
+
+    ready: function() {
+     this.updateChildren();
+    },
+
+    /**
+     * The `queryUpdated` method is the callback for whenever the
+     * `analytics-dashboard-control-change` event is fired. It updates the
+     * query attribute, which is then sent to child charts.
+     *
+     * @method queryUpdated
+     * @param {CustomEvent} event - The event with the query data.
+     */
+    queryUpdated: function(event) {
+      // Update `this.query` with the passed event data.
+      Object.keys(event.detail).forEach(function(key) {
+        this.query[key] = event.detail[key];
+      }.bind(this))
+
+      this.updateChildren();
+    },
+
+    /**
+     * The `updateChildren` method updates each of this dashboards
+     * `<google-analytics-chart>` element with its current query value.
+     *
+     * @method updateChildren
+     */
+    updateChildren: function() {
+      if (!this.authorized) return;
+      var charts = Polymer.dom(this).querySelectorAll('google-analytics-chart');
+      for (var i = 0, chart; chart = charts[i]; i++) {
+        Object.keys(this.query).forEach(function(key) {
+          chart[key] = this.query[key];
+        }.bind(this));
       }
+    },
+
+    signedIn: function() {
+      this.authorized = true;
+      this.updateChildren();
+    },
+
+    signedOut: function() {
+      this.authorized = false;
+    },
+
+    handleQueryError: function(event) {
+      this.errorMessage = event.detail;
+      // Invalid credentials after a multi-hour session have been observed in the wild.
+      this.errorMessage = "Error loading data: " + event.detail.message + " Reload might help.";
```

Comment:
The biggest problem right now is the error message doesn't go away after the error is fixed. If you think fixing that is easier than removing the code, I'd be happy to leave it in.

> no idea when this code will get looked at again

I plan to promote the use of these elements much more widely in the Google Analytics community once Polymer 1.0 is released. When we start building real-world demos that we maintain, I'll definitely be spending more time with this code.